kv/txn_coord_sender.go

```diff
@@ -336,7 +336,9 @@ func (tc *TxnCoordSender) Send(ctx context.Context, ba roachpb.BatchRequest) (*r
 					EndKey: endKey,
 				})
 			})
-			roachpb.MergeSpans(&et.IntentSpans)
+			// TODO(peter): Populate DistinctSpans on all batches, not just batches
+			// which contain an EndTransactionRequest.
+			ba.Header.DistinctSpans = roachpb.MergeSpans(&et.IntentSpans)
```

Comment:
It's subtle to rely on the fact that there has been no deduping of intent spans prior to this point. (mainly when we're merging the intents across multiple batches)

Reply:
Three years late to the party, but why does it matter whether there was deduping of intent spans prior to this point? All we care about is whether there are overlaps between requests in the current batch, right? So previous intent spans (deduped or not) may result in `MergeSpans(&et.IntentSpans)` returning false negatives where we detect overlaps between requests that don't actually exist, but never false positives where we don't detect overlaps between requests in this batch that do exist.

Am I missing something? Does that fact that this batch might overlap with previously written intents even matter past this being a convenient place to check whether the current batch contains only distinct writes?

Reply:
I just took a look and agree. `DistinctSpans` is used in a few places only: some requests use a `.Distinct()` batch if they can (`InitPut` and some others), so here it doesn't matter what happened in earlier batches (the distinct batch will see it anyway).

Then there's optimizePuts: https://github.com/cockroachdb/cockroach/blob/c5dc833a315bc4ef8fc05e9ff6180f7c92ce71de/pkg/storage/replica_evaluate.go#L33-L44

which also needs to know whether the current batch is self-overlapping. If `distinctSpans` is true, it avoids checking this again (i.e. just assumes nothing self-overlaps in the current batch) while when false, it'll actually try to do this again (to optimize at least the non-overlapping prefix of the batch). For the optimizable prefix, if it's long enough and the target keyspace is empty (verified via an iter), it switches to blind puts.

Here too it doesn't matter what happened before the current batch, it's either in the keyspace or it isn't, but ISTM that `DistinctSpans` should look at the current batch only and that's that.